### PR TITLE
Rely on gradle 4.6's annotationProcessor logic if that configuration exists

### PR DESCRIPTION
**Pull request by dansanduleac**
## Context

In gradle 4.6+, if there are any dependencies in the `annotationProcessor` configuration (added by default by the `java` plugin), then gradle automatically configures java compile tasks with `-processorpath` arguments.
There is actually one such configuration per sourceSet, so by default there is `annotationProcessor` (for `main`) and `testAnnotationProcessor` (for `test`).

## Before this PR

If the `annotationProcessor` configuration exists, we make our `processor` configuration extend from it. This conflicts with gradle's automatic handling of the `annotationProcessor` task, such that if that `annotationProcessor` is not empty, our logic (which was adding the relevant `-processorpath` flag to all JavaCompile tasks) gets superseded by gradle adding the same flag.

Therefore, dependencies in `processor` but not in `annotationProcessor` would not be picked up.

## After this PR

If the `annotationProcessor` configuration exists, this plugin no longer configures any `JavaCompile` tasks, and instead just makes all the annotationProcessor configurations for each source set extend our `processor` configuration, to ensure that processors added in older code are picked up and configured by gradle.

Additionally, we collect the `processor` AND (potentially) all `annotationProcessor` configurations into a sink configuration called `allProcessors`, and use that when configuring IDEs, to ensure that all processors will be included.

### CODE REVIEW

File: src/main/groovy/org/inferred/gradle/ProcessorsPlugin.groovy [hunk 1]
```diff
@@ -220,22 +235,6 @@ class ProcessorsPlugin implements Plugin<Project> {
     })
   }
 
-  /** Runs {@code action} on element {@code name} in {@code collection} whenever it is added. */
-  private static <T> void withName(
```
Comment (dansanduleac) [inline src/main/groovy/org/inferred/gradle/ProcessorsPlugin.groovy]: unused function

**Comment by robert3005** [comment]
LGTM

File: src/main/groovy/org/inferred/gradle/ProcessorsPlugin.groovy [hunk 2]
```diff
@@ -53,19 +94,28 @@ class ProcessorsPlugin implements Plugin<Project> {
...
       }
-    })
+    }
+  }
 
-    /**** Eclipse *********************************************************************************/
-    project.plugins.withType(EclipsePlugin, { plugin ->
-      project.plugins.withType(JavaBasePlugin, { javaBasePlugin ->
-        project.plugins.withType(JavaPlugin, { javaPlugin ->
+  private void configureEclipsePlugin(Project project, Configuration processorConf) {
```
Comment (iamdanfox) [inline src/main/groovy/org/inferred/gradle/ProcessorsPlugin.groovy]: s/processorConf/allProcessorsConf/

**Comment by iamdanfox** [comment]
I like the approach here :) I'd want to do a quick `publishToMavenLocal` to make sure this actually achieves the desired outcome before we release!

**Pull request merged**